### PR TITLE
Add support for multiple devices and start with MPPT 75|15

### PR DESCRIPTION
**Pull request by chevdor**
Hello @dbr,
I was about to start a similar project when I found your work and I think it would be nicer to work together than to have 2 projects for the same purpose :)

I started checking your code and saw that it is for now very "Bmv_700 only" :)

I started refactoring the code mainly to split devices specifics and common types / helpers and I added a new device to test the new implementation with the MPPT 75|15 that is conveniently sitting on my desk.

Before I go any further, I would like to know if that goes into a direction you like and if you would consider such contribution.
I do own a few Victron devices (MPPT, Phoenix Inverter, BMV, Battery Protect) and I would love to add support for them.

I am looking forward to hearing what you think.

### CODE REVIEW

**Comment by dbr** [comment]
This looks great, thanks! Definitely interested in having the library support all devices the protocol allows for, I was only avoiding them as it was a bit of work to fill out the appropriate fields, and I only have the BMV hooked up to test

**Comment by dbr** [comment]
Had quick look through the existing commits and looks good so far 😁

**Comment by chevdor** [comment]
Ok awesome. I will start by fully adding the MPPT 75|15 and add a few more tests.

**Comment by chevdor** [comment]
Hello @dbr, 

I think I am doing some good progresses.
I have now the functions to calculate and verify the checksum, those being also now included in the parser.
The MPPT structure should now be 98% covered (I still have a few fields mysteries to solve :)) 
I also added many tests:
```
running 35 tests
test checksum::tests_checksum::test_append_checksum ... ok
test checksum::tests_checksum::test_calculate_text_checksum_2 ... ok
test checksum::tests_checksum::test_calculate_text_checksum_for_data ... ok
test checksum::tests_checksum::test_calculate_text_checksum_real ... ok
test bmv::tests_mppt::test_mapping ... ok
test checksum::tests_checksum::test_calculate_text_checksum_short ... ok
test checksum::tests_checksum::test_calculate_text_checksum_short2 ... ok
test checksum::tests_checksum::test_verify_text_checksum_2 ... ok
test checksum::tests_checksum::test_verify_text_checksum_real ... ok
test checksum::tests_checksum::test_verify_text_checksum_real_error ... ok
test frames::mppt_frame::tests_mppt::test_mppt_init ... ok
test frames::mppt_frame::tests_mppt::test_mppt_to_bytes ... ok
test frames::mppt_frame::tests_mppt::test_mppt_1 ... ok
test frames::mppt_frame::tests_mppt::test_mppt_new ... ok
test frames::mppt_frame::tests_mppt::test_mppt_to_string ... ok
test frames::mppt_frame::tests_mppt::test_mppt_older_versions ... ok
test parser::test_frame_finder::test_find_frame_burried_frame ... ok
test parser::test_frame_finder::test_find_frame_frame_then_junk ... ok
test parser::test_frame_finder::test_find_frame_junk_then_frame ... ok
test parser::test_frame_finder::test_find_frame_noframe ... ok
test parser::test_frame_finder::test_findstart_none ... ok
test parser::test_frame_finder::test_find_frame_only_frame ... ok
test parser::test_frame_finder::test_findstart_some ... ok
test parser::test_frame_finder::test_findstart_some_0 ... ok
test parser::test_frame_finder::test_truncate_nothing ... ok
test parser::test_frame_finder::test_truncate_some ... ok
test parser::tests_parser::test_parse_hex ... ok
test parser::tests_parser::test_parse_line ... ok
test parser::tests_parser::test_parse_nonsense ... ignored
test parser::tests_parser::test_parse_serial ... ok
test parser::tests_parser::test_partial_stream ... FAILED
test utils::tests_utils::test_convert_percentage2 ... ok
test utils::tests_utils::test_convert_percentage ... ok
test utils::tests_utils::test_get_field_string_none ... ok
test utils::tests_utils::test_get_field_string_some ... ok
```

There are a few points I would like to discuss with you.

## Streaming
I saw, running the tests on master, that a test is not passing:
```
...
test parser::tests_parser::test_partial_stream ... FAILED
```

I don't think it can pass with the parser as it is and I started working in a direction to support such a streaming as we typically get from a serial port. 

I was wondering however if you had already plans and/or ideas about how to address the issue.
I saw that you avoided the issue in the example with a properly timed `sleep` :)

## Suspicious test

In the following test: https://github.com/dbr/vedirect-rs/blob/master/src/data.rs#L120, you pass `...V\t232...`.
According to the specs, that should give us `0.232V` (or 232 mV) and not `23.2V` as mentioned in the test. 
Could it be an error or does the BMV-700 really behaves like that?

## Checksum

You defined `Checksum` as a struct (https://github.com/dbr/vedirect-rs/blob/master/src/parser.rs#L10) with a single `value` fields. Did you have a reason for now using a simple type alias there ?

**Comment by dbr** [comment]
That looks like excellent progress!

> I don't think it can pass with the parser as it is and I started working in a direction to support such a streaming as we typically get from a serial port.

This was one of the parts I got stuck on, how best to structure the nom parser to reliably deal with starting in the middle of streams etc. I think that failing test was to remind me to improve the parser and remove the need for the sleep hackery!

Feel free to improve the parser to handle this however you think best. I think I was considering just writing the parser by hand instead of using nom, since the protocol is so simple

It would also be nice to remove nom as a dependency if it doesn't provide enough value - as I imagine the library will often be used on low-power devices like Raspberry Pi's etc, so reducing the number of crates and macros etc will help with compile times if not cross-compiling

> Could it be an error or does the BMV-700 really behaves like that?

Very likely an error on my part - not too sure why I ended up dividing it by 10 - the HEX protocol doc has the voltages in `0.01V` units, and the text protocol in mV.. and neither of those are tenth of a volt!

Probably worth noting I haven't used the library heavily myself yet - it was written to basically just record historical power usage and SoC in a campervan I had finished building earlier this year, but since I've haven't traveled much in it lately, I hadn't finished the logger.. thus meaning I haven't closely looked at any of the data yet - so definitely trust the spec over my existing code!

> You defined Checksum as a struct (https://github.com/dbr/vedirect-rs/blob/master/src/parser.rs#L10) with a single value fields. Did you have a reason for now using a simple type alias there ?

I think just for consistency with the `Field` struct, and so when printed with things like `dbg!(...)` it would be a bit more explicit (e.g `Checksum { value: 123 }` instead of just `123`). I think I was half-planning to put the checksum verification logic as an associated method with that struct too, although that may not make much sense in practice

File: src/bmv.rs [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+    pub soc: Option<Percent>,
+
+    /// Time-to-go. Labelled `TTG`
+    /// Units: Minutes (When the battery is not discharging the time-to-go is infinite. This is represented as -1)
+    /// Available on: BMV 600, BMV 700
+    pub ttg: Minute,
+}
+
+impl Map<Bmv700> for Bmv700 {
+    /// "When the BMV is not synchronised, these statistics have no meaning, so "---" will be sent instead of a value"
```
Comment (dbr) [inline src/bmv.rs]: Minor thing: misplaced comment

File: src/checksum.rs [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+// }
+
+pub type Checksum = u8;
+
+/// Calculate the checksum of some data
+pub fn calculate(data: &[u8]) -> Checksum {
+    let mut checksum = 0u8;
+
+    data.iter().for_each(|x| {
+        checksum = (checksum.overflowing_add(*x)).0.modulo(255);
```
Comment (dbr) [inline src/checksum.rs]: I could very well be mistaken, but: could this use the built in remainder `%` operator instead of the external modulo method? For an unsigned integer, [they should be equivalent I think](https://stackoverflow.com/a/13683709)

Comment (chevdor) [reply src/checksum.rs]: the modulo crate is actually super straight forward... Sure we could replace it by our own implementation but that would be more of a copy/paste vs using a crate. There will not be much of a benefit IMO.
Once we have more tests and benchmarks though, those are some tests we can run to see if there is a benefit.

**Comment by chevdor** [comment]
Thanks for your answer and feedback @dbr.
As you spotted, there is still some cleanup to do :)

> This was one of the parts I got stuck on, how best to structure the nom parser to reliably deal with starting in the middle of streams etc.<...>
> Feel free to improve the parser to handle this however you think best. I think I was considering just writing the parser by hand instead of using nom, since the protocol is so simple

Ok, cool. I started going this way. For now, I did keep your parser as it works pretty good. I started adding helping functions you may have seen. They should allow handling streaming.

> It would also be nice to remove nom as a dependency if it doesn't provide enough value - as I imagine the library will often be used on low-power devices like Raspberry Pi's etc, so reducing the number of crates and macros etc will help with compile times if not cross-compiling

Running on small devices should definitely be the goal. IMO, the number of crates or compile time should not be a concern. However, the number of CPU cycles and memory used should be for small / embedded devices. As far as I can tell, `nom` is doing pretty good there. I do not plan on removing it now to keep the PR small (well... not too huge at least :)). Once things work, we can still revisit and do some more testing. I agree that nom should be replaceable if there is a benefit. With the tests I am bring, that should be pretty safe later on.

> Very likely an error on my part - not too sure why I ended up dividing it by 10 - the HEX protocol doc has the voltages in 0.01V units, and the text protocol in mV.. and neither of those are tenth of a volt!

OK, no worries. I actually have a BMV-712, I will test it as well once I am done with the rest.

> (checksum) I think just for consistency with the Field struct, and so when printed with things like dbg!(...) it would be a bit more explicit (e.g Checksum { value: 123 } instead of just 123). I think I was half-planning to put the checksum verification logic as an associated method with that struct too, although that may not make much sense in practice

Yes, indeed, I was considering that but none of the options convinced me so far. If we need, this is rather easy to change.

File: src/checksum.rs (hunk 2 above)
Comment (dbr) [reply src/checksum.rs]: Oh it's definitely not worth implementing the crate, I'm just wondering if

    use modulo::Mod;
    checksum = (checksum.overflowing_add(*x)).0.modulo(255);

could be replaced with [`std::op::Rem`](https://doc.rust-lang.org/std/ops/trait.Rem.html) in this case, e.g

    checksum = (checksum.overflowing_add(*x)).0 % 255;

Comment (chevdor) [reply src/checksum.rs]: So I ran some tests and benchmarks with the code below.

<details>
<summary>tests & benchmarks</summary>
    #[test]
    fn test_checksum_crate_vs_alt() {
        let data = vec![0x0d, 0x0a, 0xf0, 0x0f, 0xff, 0x55];
        let checksum1 = calculate(&data);
        let checksum2 = calculate_alt(&data);
        assert_eq!(checksum1, checksum2);
    }

    #[bench]
    fn test_checksum_crate(b: &mut Bencher) {
        let data = vec![0x0d, 0x0a, 0xf0, 0x0f, 0xff, 0x55];
        let checksum = calculate(&data);
        assert_eq!(checksum, 150);
        b.iter(|| calculate(&data));
    }

    #[bench]
    fn test_checksum_no_crate(b: &mut Bencher) {
        let data = vec![0x0d, 0x0a, 0xf0, 0x0f, 0xff, 0x55];
        let checksum = calculate_alt(&data);
        assert_eq!(checksum, 150);
        b.iter(|| calculate(&data));
    }
</details>



The result of the benchmarks is:
```
test checksum::benchmarkss::test_checksum_crate    ... bench:           6 ns/iter (+/- 0)
test checksum::benchmarkss::test_checksum_no_crate ... bench:           6 ns/iter (+/- 0)
```

So in short:
- it works just as well without the extra crate
- the timing is the same in both cases

As conclusion, I don't think it makes sense using an extra crate when we don't need to. I will remove it.

**Comment by felixwatts** [comment]
What happened here?
